### PR TITLE
fallback-scsv: add support for DHE/ECDHE ciphersuites

### PR DESCRIPTION
**Pull request by tomato42**
### Description

add support for DHE/ECDHE ciphers for FIPS mode compatibility

### Motivation and Context

#563 

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [x] the changes are also reflected in documentation and code comments
- [ ] all new and existing tests pass (see Travis CI results)
- [ ] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [x] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [x] new and modified scripts were ran against popular TLS implementations:
  - [x] OpenSSL
  - [ ] NSS
  - [ ] GnuTLS
- [x] required version of tlslite-ng updated in requirements.txt and README.md

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/574)

### CODE REVIEW

File: scripts/test-fallback-scsv.py [hunk 1]
```diff
@@ -69,23 +75,42 @@ def main():
...
+            SignatureAlgorithmsCertExtension().create(RSA_SIG_ALL)
+        ciphers = [CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
+                   CipherSuite.TLS_DHE_RSA_WITH_AES_128_CBC_SHA,
+                   CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV,
+                   CipherSuite.TLS_AES_128_GCM_SHA256]
+    else:
+        ext = None
+        ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+                   CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV,
+                   CipherSuite.TLS_AES_128_GCM_SHA256]
```
Comment (nmav) [inline scripts/test-fallback-scsv.py]: This looks like a TLS1.3 ciphersuite but no TLS1.3 seems to be enabled. Is its addition here intentional?

File: scripts/test-fallback-scsv.py [hunk 2]
```diff
@@ -36,24 +38,28 @@ def help_msg():
...
-    opts, args = getopt.getopt(argv, "h:p:e:", ["help", "tls-1.3"])
+    opts, args = getopt.getopt(argv, "h:p:e:d", ["help", "tls-1.3"])
     for opt, arg in opts:
         if opt == '-h':
             host = arg
         elif opt == '-p':
             port = int(arg)
         elif opt == '-e':
             run_exclude.add(arg)
+        elif opt == '-d':
```
Comment (nmav) [inline scripts/test-fallback-scsv.py]: Why test it with a flag and not with a separate test? Also seeing below ECDHE and DHE cannot be tested separately, ECDHE will only be tested by a server that supports both.

File: scripts/test-fallback-scsv.py [hunk 3]
```diff
@@ -211,21 +326,45 @@ def main():
...
+        positions = (0, 1, 2, 3)
+    else:
+        positions = (0, 1, 2)
+    for place in positions:
         conversation = Connect(host, port)
         node = conversation
-        ciphers = [CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV,
-                   CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
-                   CipherSuite.TLS_AES_128_GCM_SHA256]
+        if dhe:
```
Comment (nmav) [inline scripts/test-fallback-scsv.py]: It is necessary to have this test both in DHE and RSA? It looks like what's it is testing (rejection in position) doesn't depend on what ciphersuite is negotiated in the end.

**Comment by tomato42** [comment]
> Overall it looks straightforward, no obvious issues. However, I am not sure whether all the added checks were necessary, not whether a separate run of the script is needed to check DHE. Since there is no description in the commit msg about this advising otherwise, I'd have expected new tests to be added instead, similarly to other test scripts.

see #563 – the idea is to be able to run the scripts against servers that cannot enable RSA key exchange

File: scripts/test-fallback-scsv.py (hunk 1 above)
Comment (tomato42) [reply scripts/test-fallback-scsv.py]: yes, it's to check that the server does not use presence of TLS 1.3 ciphersuites as a proxy for supported_versions with TLS 1.3

**Comment by nmav** [comment]
> see #563 – the idea is to be able to run the scripts against servers that cannot enable RSA key exchange

Would it make sense to include that issue and the description in the commit message? It would clarify the purpose and the reasons for this change.

**Comment by tomato42** [comment]
> > see #563 – the idea is to be able to run the scripts against servers that cannot enable RSA key exchange
> 
> Would it make sense to include that issue and the description in the commit message? It would clarify the purpose and the reasons for this change.

added, though finding on github the PR that the commit was introduced in is quite easy, so usefulness of that is a bit reduced...

**Pull request merged**